internal/shim-sev/src/syscall.rs

```diff
@@ -112,778 +112,290 @@ extern "sysv64" fn syscall_rust(
-                    );
-                    return Err(libc::EINVAL);
-                }
-            }
+    fn trace(&mut self, name: &str, argc: usize) {
+        eprint!("{}(", name);
+        for (i, arg) in self.argv[..argc].iter().copied().enumerate() {
+            let prefix = if i > 0 { ", " } else { "" };
+            eprint!("{}0x{:x}", prefix, arg);
         }
+
+        eprintln!(")");
     }
-    eprintln!("SC> mprotect({:#?}, {}, {}, …) = 0", addr, len, prot);
 
-    Ok(0)
-}
+    fn exit(&mut self, status: i32) -> ! {
+        self.trace("exit", 1);
+        hostcall::shim_exit(status);
+    }
 
-/// syscall
-pub fn mmap(
-    addr: *const u8,
-    length: libc::size_t,
-    prot: libc::c_int,
-    flags: libc::c_int,
-    fd: libc::c_int,
-    offset: libc::off_t,
-) -> Result<usize, libc::c_int> {
-    const PA: i32 = libc::MAP_PRIVATE | libc::MAP_ANONYMOUS;
-
-    match (addr as u64, length, prot, flags, fd, offset) {
-        (0, _, _, PA, -1, 0) => {
-            let mut flags = PageTableFlags::PRESENT | PageTableFlags::USER_ACCESSIBLE;
-
-            if prot & libc::PROT_WRITE != 0 {
-                flags |= PageTableFlags::WRITABLE;
-            }
+    fn exit_group(&mut self, status: i32) -> ! {
+        self.trace("exit_group", 1);
+        hostcall::shim_exit(status);
+    }
 
-            if prot & libc::PROT_EXEC == 0 {
-                flags |= PageTableFlags::NO_EXECUTE;
+    fn arch_prctl(&mut self, code: i32, addr: u64) -> sallyport::Result {
+        self.trace("arch_prctl", 2);
+        match code {
+            ARCH_SET_FS => {
+                unsafe {
+                    wrfsbase(addr);
+                }
+                eprintln!("SC> arch_prctl(ARCH_SET_FS, {:#x}) = 0", addr);
+                Ok(Default::default())
             }
-
-            let virt_addr = *NEXT_MMAP_RWLOCK.read().deref();
-            let len_aligned = align_up(length as _, Page::<Size4KiB>::SIZE) as _;
-
-            let mem_slice = FRAME_ALLOCATOR
-                .write()
-                .allocate_and_map_memory(
-                    SHIM_PAGETABLE.write().deref_mut(),
-                    virt_addr,
-                    len_aligned,
-                    flags,
-                    PageTableFlags::PRESENT
-                        | PageTableFlags::WRITABLE
-                        | PageTableFlags::USER_ACCESSIBLE,
-                )
-                .map_err(|_| {
-                    eprintln!("SC> mmap({:#?}, {}, …) = ENOMEM", addr, length);
-                    libc::ENOMEM
-                })?;
-            eprintln!(
-                "SC> mmap({:#?}, {}, …) = {:#?}",
-                addr,
-                length,
-                mem_slice.as_ptr()
-            );
-            unsafe {
-                core::ptr::write_bytes(mem_slice.as_mut_ptr(), 0, length);
+            ARCH_GET_FS => unimplemented!(),
+            ARCH_SET_GS => unimplemented!(),
+            ARCH_GET_GS => unimplemented!(),
```

Comment:
Should arch_prctl just return ENOSYS here?

Reply:
just implemented them in a separate patch